Enable restore for ref and src projects in libraries

ViktorHofer opened the pull request:
Fixes https://github.com/dotnet/runtime/issues/29953
Contributes towards https://github.com/dotnet/runtime/issues/31844

Upfront restore is now taking far too long, presumably because of project evaluation or multiple globs. I'm considering making restore implicit (during the build of the refs, sources and tests) so that the initial restore only handles the depproj files. Under a switch the restore target would still be able to upfront restore everything (CI mode).

Later when we removed all depproj files and sequenced _everything_ into the build, we can remove the `-restore` action which is passed to the Arcade's Build.proj so that the `-restore` action is only invoked on demand (CI or before going offline, ie airplane).

### Requires
- [x] https://github.com/dotnet/arcade/pull/4989
- [x] https://github.com/dotnet/arcade/pull/4992
- [x] https://github.com/dotnet/arcade/pull/4993
- [x] https://github.com/dotnet/arcade/pull/4996
- [ ] https://github.com/dotnet/sdk/pull/10794

cc @Anipik @joperezr @safern

ViktorHofer committed:
Enable restore for projects in libraries

Enabling project restore for ref and src projects
(any by that any others as well).

ViktorHofer committed:
Remove winrt.depproj

ViktorHofer committed:
Remove tools.depproj

ViktorHofer committed:
Enable upfront restore

ViktorHofer committed:
Stop flowing of package dependencies

joperezr commented on eng/Analyzers.props:
```diff
@@ -1,6 +1,8 @@
   <PropertyGroup>
     <CodeAnalysisRuleset>$(MSBuildThisFileDirectory)CodeAnalysis.ruleset</CodeAnalysisRuleset>
+    <!-- Disable using analyzers in sourcebuild -->
+    <EnableAnalyzers Condition="'$(DotNetBuildFromSource)' == 'true'">false</EnableAnalyzers>
```
This doesn't seem to be related to this change right? perhaps we should separate it on a different PR?

ViktorHofer replied:
This is related, as before src and ref assemblies were going through the manual target that @safern wrote which adds the Analyzers to the csc task. Now they go through the same path as test assemblies and other libs like CoreLib. I moved this out from Tools.props and applied the condition here for sourcebuild.

joperezr commented on eng/Tools.props:
```diff
@@ -1,45 +1,15 @@
 <Project>
-  <PropertyGroup>
-    <MSBuildTreatWarningsAsErrors>false</MSBuildTreatWarningsAsErrors>
```
We probably still want the TreatWarningsAsErrors right? In order to catch errors where one of the packages you were trying to restore wasn't available so you got a different one instead.

ViktorHofer replied:
Right but this was set to false before

joperezr replied:
Ohh right! haha my bad.

joperezr commented on eng/Tools.props:
```diff
@@ -1,45 +1,15 @@
 
   <!-- excluded from source build -->
   <ItemGroup Condition="'$(DotNetBuildFromSource)' != 'true'">
-    <PackageReference Include="Microsoft.DotNet.ApiCompat" Version="$(MicrosoftDotNetApiCompatVersion)" />
```
Where will we be getting all of these references now? I haven't found where all of these tools reference went, same for the compilers toolset and Genfacades.

joperezr replied:
Oh nvm, I just found them 😄

ViktorHofer replied:
https://github.com/dotnet/runtime/pull/33242/files#diff-d3bedba39ce6da3e37ca0db83d7ef035R159-R164, and for the compiler https://github.com/dotnet/arcade/blob/master/src/Microsoft.DotNet.Arcade.Sdk/tools/Compiler.props#L5.

ViktorHofer committed:
Merge illink.tasks packageref with others

ViktorHofer committed:
Update TargetFramework.Sdk

joperezr commented on src/libraries/ref.proj:
```diff
@@ -0,0 +1,36 @@
+<Project InitialTargets="IncludeProjectReferences">
```
NIT: Looks like git is not considering this and ref.builds to have been the same file, which sometimes breaks the ability to check history on the file or to blame a line. When making these renames, can we try using `git mv oldFileName newFileName` so that history is preserved?

joperezr replied:
It also makes it easier to review 😉

ViktorHofer replied:
just did that, thanks

joperezr commented on src/libraries/ref.proj:
```diff
@@ -0,0 +1,36 @@
+             BuildInParallel="true"
+             ContinueOnError="ErrorAndStop" />
+
+    <!-- Given we ErrorAndContinue we need to propagate the error if the overall task failed -->
```
Looks from above that now we have it as ErrorAndStop, so we probably don't need this last Error task any more?

ViktorHofer replied:
I believe in the past I copied that block from dir.traversal.targets. Let's try it out.

ViktorHofer replied:
Oh I see, for testing it is ErrorAndContinue which makes sense.

joperezr commented on src/libraries/src.proj:
```diff
@@ -17,5 +21,23 @@
+
+  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.targets" />
+  
+  <Target Name="Build" DependsOnTargets="BuildAllProjects" />
```
It might be worth it to add a comment above Build target to say that ordering matters and that we are intentionally defining this after importing sdk.targets to override that value, that way if people in the future edit this file will keep that in mind.

joperezr commented on src/libraries/src.proj:
```diff
@@ -1,5 +1,9 @@
+  <Import Sdk="Microsoft.NET.Sdk" Project="Sdk.props" />
+
+  <PropertyGroup>
+    <TargetFramework>netcoreapp5.0</TargetFramework>
```
$(NetCoreAppCurrent) instead?

ViktorHofer replied:
This one is problematic. I believe we need to set a TargetFramework here so that NuGet doesn't complain. But I'm not sure if this TargetFramework has any impact on the chosen projects. I don't know what the right way is here, maybe setting it to BuildTargetFramework? cc @Anipik 

This needs to be solved for ref.proj, src.proj and tests.proj.

ViktorHofer committed:
Error task cleanup in composits

ViktorHofer committed:
Add comments for Target ordering

ViktorHofer committed:
Set composit tfms to BuildTargetFramework

ViktorHofer committed:
Disable project restore for others

ViktorHofer committed:
Fix analyzers and build references in P2P

ViktorHofer commented on src/libraries/Directory.Build.props:
```diff
@@ -351,6 +352,12 @@
+    <EnableProjectRestore Condition="'$(IsSourceProject)' == 'true' or '$(IsReferenceAssembly)' == 'true' or '$(IsTestProject)' == 'true'">true</EnableProjectRestore>
+  </PropertyGroup>
+
+  <Import Condition="'$(EnableProjectRestore)' != 'true'" Project="$(RepositoryEngineeringDir)restore\repoRestore.props" />
```
- depproj (which doesn't really count, as we never invoke the restore target as entrypoints on them)
- pkgproj

ericstj replied:
I see, we should try to clean up this mode, even if it means doing something slightly different for these.  I was hoping to completely eliminate this hack.

ViktorHofer commented on src/libraries/src.proj:
```diff
@@ -17,5 +21,21 @@
-  <Import Project="$(MSBuildThisFileDirectory)dir.traversal.targets" />
+  <Target Name="IncludeProjectReferences">
+    <ItemGroup>
+      <ProjectReference Include="@(Project)" />
```
Before with the old configuration system and the FilterProjects target, this was necessary (in tests.proj).

ViktorHofer committed:
Reenable codeanalysis package

ViktorHofer committed:
Fix casing of PInvokeAnalyzerException file

ViktorHofer committed:
Update global.json

ViktorHofer committed:
Update Versions.props

ViktorHofer committed:
Update Version.Details.xml

ViktorHofer committed:
Enable static graph evaluation for restore

Anipik commented on src/libraries/build.proj:
```diff
@@ -19,10 +19,12 @@
+      <_RestoreProjects Include="$(MSBuildThisFileDirectory)dirs.proj" Build="false" />
+      <_RestoreProjects Include="$(MSBuildThisFileDirectory)tests.proj" Condition="'$(DotNetBuildFromSource)' != 'true'" Build="false" />
     </ItemGroup>
+    
```
nit : extra line

Anipik commented on src/libraries/dirs.proj:
```diff
@@ -1,26 +1,37 @@
   </PropertyGroup>
 
   <ItemGroup Condition="'$(DirectoryToBuild)' == ''">
-    <Project Include="ref.builds" />
```
just for curiosity, why are we converting .build to projs ?

joperezr replied:
Yes, but he wants to control the exact time where the Sdk.targets get loaded, so he can then override the Build target. Check bellow where he actually added a comment about this.

ViktorHofer replied:
For Intellisense to work by default. .*proj is also restorable by nuget.

Anipik commented on src/libraries/dirs.proj:
```diff
@@ -1,26 +1,37 @@
 <Project>
```
isnt adding ```<Project Sdk="Microsoft.NET.Sdk">``` will import both the sdk.props and sdk.targets ?

ViktorHofer committed:
Make ApiCmpat.proj restorable

ViktorHofer committed:
Optimize restore and disable RestoreTaskEx meanwhile

ViktorHofer committed:
Remove dir.traversal.tarets from depproj container

ViktorHofer committed:
Remove harvestPackages.depproj and refactoring

Removing havestPackages.depproj in favor of a target that is hooked
into Arcade's repo restore phase in Tools.props.

Additionally moving common properties that are used both in Arcade and
the repo inside Configurations.props. Before they were scattered around
multiple places (Versions.props, Subsets.props).

ViktorHofer committed:
Fix WindowsRuntime test

ViktorHofer committed:
Revert dirs.proj depproj change

ViktorHofer committed:
Add sourcelink back to Tools.props

ViktorHofer committed:
Set ilasm/ildasm RID for restore

ViktorHofer commented:
@ericstj this is ready. PTAL

ViktorHofer committed:
Sequence restore into steps out of upfront restore

ericstj commented on src/libraries/dirs.proj:
```diff
@@ -1,26 +1,37 @@
-    <Project Include="shims\ApiCompat.proj" Condition="'$(DotNetBuildFromSource)' != 'true' and '$(BuildTargetFramework)' == '$(NetCoreAppCurrent)'" />
-    <Project Include="pretest.proj" Condition="'$(DotNetBuildFromSource)' != 'true' and '$(BuildAllConfigurations)' != 'true'" />
+    <ProjectReference Include="ref.proj" />
+    <ProjectReference Include="shims\generated\*.csproj" />
```
Can you move this into ref.proj and build as another target after the main build?  That would make ref.proj more useful, even though it does make it slightly more complicated.

ViktorHofer replied:
If you don't feel strongly about this I would prefer to not do so as we will delete these files (ref.proj, src.proj, tests.proj, dirs.proj, build.proj, package.builds) anyway as we want to define subsets and add the projects as ProjectToBuild arcade extension point items.

ericstj replied:
Sure, don't do anything that's throw away, but understand that dirs.proj is representing key sequencing dependencies, so I suspect these things will need to continue to be grouped somehow.  It'd be nice if they were grouped in a way that permitted folks to perform the common task of "build all inputs needed for building src projects" which is refs and shims.

ViktorHofer replied:
I was already in that file so I did it anyway :)

Yeah that's possible with the subsets. We can follow-up on that later in a few weeks.

ericstj commented on tools-local/tasks/installer.tasks/installer.tasks.csproj:
```diff
@@ -20,6 +20,14 @@
 
+  <!--
+    Restore the packaging package here to use it during the repo restore phase in 
+    harvestPackages.targets which is hooked into Tools.props.
```
Interesting hack.  I don't suppose there was a cleaner way to get this restored earlier?

ViktorHofer replied:
Unfortunately not unless we bring the necessary task into the repo or sequence the harvesting package download out of the repo restore phase.

ericstj replied:
I see.  Do our Proj files support restore?  If so we could have repo restore download the package and packages.proj do the PackageDownloads.

ViktorHofer replied:
That's a great idea. Yeah *proj files are restorable by default since a nuget update (I believe last year).

ViktorHofer committed:
Fix docker entrypoints by using build

ViktorHofer committed:
Move build of generate shims into ref.proj

ViktorHofer committed:
PR feedback

ViktorHofer committed:
Fixing enterprise pipeline

ViktorHofer committed:
Revert dirs.proj depproj refactoring

ViktorHofer committed:
Enable restore in dirs.proj libraries

ViktorHofer committed:
It must work!!

ViktorHofer committed:
Please work

joperezr commented:
![image](https://user-images.githubusercontent.com/13854455/76259203-5d272000-6212-11ea-86bd-cda7f181fd2f.png)


@ViktorHofer 😆 🤣

jaredpar commented:
Oh man I laughed so hard when I read this it scared the dog

![image](https://user-images.githubusercontent.com/146967/76259231-6ca66900-6212-11ea-9504-f63294b32fd9.png)

ViktorHofer committed:
Fix one error at a time

ViktorHofer committed:
Add packages.proj to upfront restore

ViktorHofer committed:
Fix sequencing

ViktorHofer committed:
Remove unnecessary packageref

ViktorHofer committed:
More fixes in ProjectRef composition

ViktorHofer committed:
Undo commented out lines in dirs.proj

ViktorHofer committed:
Sequence refs correctly and cleanup

ViktorHofer committed:
Cleanup proj containers

ViktorHofer committed:
Remove wrong condition

ViktorHofer commented:
Failure is https://github.com/dotnet/runtime/issues/13295
